zuse: adds |base16 en/decoding core

joemfb opened the pull request:
Motivated by #1892, jets incoming. This is just typical MSB-first hex; the only remotely interesting part is the careful handling of input length to account for leading zeros. (The input will be `$octs` once #3554 is in.)

The API is (mostly) taken from `lib/base64`, which I intend to clean up and move into %zuse (and jet) in a similar manner.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
+            [(sub p.a wid) q.a]
+          =/  dat  (cut 3 [0 p.a] q.a)
+          =.  wid  (met 3 dat)
+          [(sub p.a wid) dat]
```
I think this should eventually be factored out into a standard helper, so we can apply this pattern consistently whenever we're rendering `$octs`.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
+          %+  cook
+            |=  waq=(list @)
+            =/  wid  (mul (lent waq) (dec (xeb wuc)))
+            :-  (rsh 0 boq (add wid (dec (bex boq))))
```
This is just standard bloq-length conversion, and is actually just `div-ceil`. It'd be nice to have bloq-length conversion in the stdlib somewhere, but I haven't worked out the details.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
+            |=  waq=(list @)
+            =/  wid  (mul (lent waq) (dec (xeb wuc)))
+            :-  (rsh 0 boq (add wid (dec (bex boq))))
+            (roll waq |=([p=@ q=@] (add (end 0 wuc p) (mul wuc q))))
```
Note that this fold is slightly different from `+bass` in hoon.hoon -- the `(end 0 wuc ...)` addition makes this a consistent bitwise operation, and much more suitable for jetting. I think it also captures the behavior we actually want here, and the version in hoon.hoon (along with `+boss`) should be probably be updated to match.

Note also that this is equivalent to `(repn wuc waq)`. I didn't use that because a) I'm jetting the decoding arm directly and b), I want this pattern to be used more generally, but the `+repn` jet as implemented is not suitable for large inputs.

ohAitch replied:
Would there be any sense in improving the `repn` jet directly?

joemfb replied:
Definitely, I plan to do so.

`u3r_chop()` and the atom slab allocator show up as the hotspots in parsers like this -- the former needs to be rewritten for efficiency, and the latter can be solved by either special-casing direct-atom sizes in all the jets, or designing a slab-allocator api that can do so automatically. I aim to try for the latter.

Fang- replied:
Given the above, this probably needs an `XX  use repn when ...`.

joemfb replied:
This code was wrong, as were my comments: my assumptions only held if the base was a power of two. I've removed the generic code here and just written what's need for base16. I'll save the more general stuff for a better time/place.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
       ?:  =(~ t.myn)  (trip i.myn)
       (weld (trip i.myn) `tape`['/' $(myn t.myn)])
+    ::
+    ::  |base16: en/decode arbitrary MSB-first hex strings
```
I went with MSB-first because that's standard for hex, and matches the behavior of our existing use-cases. But the byte-order remains an open question for the other standard encodings I'd like to see here: base32/58/64.

Our implementations have generally only been suitable for cords, and, as such, have tended to use LSB-first order. I see these conversions in zuse as being primarily about interop with external systems (otherwise the auras would work). It really just comes down to which types of input need to be byte-swapped, but it'd be nice to have consistent answers.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
+      |%
+      ++  en
+        ~/  %en
+        |=  a=(pair @ud @)  ::  $octs, but safe
```
`$octs` is actually defined in multiple places, and referenced from each. I'll open a separate PR to clean that up.

Fang- replied:
It's not immediately clear to me how that stops you from using `octs` here. There's no conflicting definition around, right?

Other than that, changes look good!

ohAitch replied:
I'm also not sure how @t is "unsafe" here tbh, like, does it not print bad bytes `\` escaped? That sounds like a printer bug, auras are always only advisory

joemfb replied:
The specific crash that motivated this may be a bug, but it's at least a deliberate design bug (`+wood`, cord-escape, calls `+teff`, which includes a bunch of assertions). I could see arguing either way in the printer.

But this data structure exists solely for the purpose of representing octet-streams that cannot be losslessly encoded as atoms, so by definition, they cannot be cords. And I think it's fine for operations on atoms with auras that have a limited range to assume that their arguments are in that range. It would certainly be a pain to check such invariants anytime one handled such an atom. As we've discussed before, I'd like to see such validation performed more vigorously, but pushed to the edge of an input/type domain (by finishing `+sane` and enabling it in atom molds).

ohAitch replied:
Hm, am I wildly off base that the vast majority of `octs` constructed are in fact cords, whether or not they cannot be? Like, octs exists for representing octet streams we *cannot rule out* having trailing zeroes - I suppose my intuition here boils down to, if you `+curl` four textual urls and one `.gif`, the resulting type should both correctly represent the full range of the latter for `can` or fs sync or w/e, and print the common case as text and not as a decimal number.

joemfb replied:
As far as I can tell, there's  currently two main use-cases for `$octs` or similar structures: http request/response bodies (in the latter case, possibly partial), and compatibility focused APIs like crypto functions (particularly hashes) and string-format en/decoding. It wouldn't surprise me if the majority of inputs in both cases are cords, but, even if so, the fact is incidental.

HTTP requests are now almost always routed to apps, where they're parsed/decoded and can be printed in an ad-hoc manner. Similarly, samples/products of various stdlib gates can be easily printed ad-hoc.

`+curl` is gone, but `+url` in still works in `:dojo`. That's the only use-case I can think of were we need to print arbitrary `$octs`. It'd be pretty easy to add a cast to cord there, gated by `(sane %t)`.

joemfb replied:
Actually, that already happens -- see `+dy-print` in `:dojo`.

joemfb commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -6036,6 +6037,57 @@
+        |=  a=@C  ^-  cord
+        ?:  (lth a 10)
+          (add '0' a)
+        (add 'a' (sub a 10))
```
Yes, great suggestion! (The minimum length needs to be multiplied by 2.)

joemfb commented:
@Fang-, I rebased this onto `release/next-sys` and updated the pill, but forgot to use `$octs` first. I'll push a new commit/pill on top, and squash if everything passes. PTAL.

Fang- reviewed (approved):
Changes look good!
